src/firstboot/firstboot.c

```diff
@@ -1049,6 +1051,75 @@ static int process_kernel_cmdline(int rfd) {
         return 0;
 }
 
+static int process_reset(int rfd) {
+        _cleanup_free_ char *buf = NULL;
+        _cleanup_close_ int esp_fd = -EBADF, loader_fd = -EBADF;
+        int r;
+
+        if (!arg_reset)
+                return 0;
+
+        FOREACH_STRING(s,
+                       "etc/locale.conf",
+                       "etc/vconsole.conf",
+                       "etc/localtime",
+                       "etc/hostname",
+                       "etc/machine-id",
+                       "etc/kernel/cmdline",
+                       "etc/machine-info",
+                       "var/lib/systemd/random-seed",
+                       "var/lib/systemd/credential.secret") {
+                _cleanup_free_ char *bn = NULL;
+                _cleanup_close_ int pfd = -EBADF;
+
+                r = chase_symlinks_at(rfd, s, CHASE_AT_RESOLVE_IN_ROOT|CHASE_PARENT, NULL, &pfd);
+                if (r == -ENOENT)
+                        continue;
+                if (r < 0)
+                        return log_error_errno(r, "Failed to resolve %s/%s: %m", FORMAT_FD_PATH(rfd, &buf), s);
+
+                r = path_extract_filename(s, &bn);
+                if (r < 0)
+                        return r;
+
+                r = RET_NERRNO(unlinkat(pfd, bn, 0));
+                if (r == -ENOENT)
+                        continue;
+                if (r < 0)
+                        return log_error_errno(r, "Failed to remove %s/%s: %m", FORMAT_FD_PATH(rfd, &buf), s);
+
+                log_info("Removed %s/%s.", FORMAT_FD_PATH(pfd, &buf), bn);
+        }
+
+        r = find_esp_at_and_warn(rfd, NULL, /* unprivileged_mode= */ false, NULL,
+                                 &esp_fd, NULL, NULL, NULL, NULL, NULL);
+        if (r < 0) {
+                if (r != -ENOKEY)
+                        return r; /* find_esp_and_warn() already logged */
+
+                log_debug_errno(r, "Couldn't find any ESP, so not removing ESP random seed.");
+                return 0;
+        }
+
+        r = chase_symlinks_at(esp_fd, "loader", CHASE_AT_RESOLVE_IN_ROOT|CHASE_PROHIBIT_SYMLINKS,
+                              NULL, &loader_fd);
+        if (r == -ENOENT)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to resolve %s/loader: %m", FORMAT_FD_PATH(esp_fd, &buf));
+
+        r = RET_NERRNO(unlinkat(loader_fd, "random-seed", 0));
+        if (r == -ENOENT)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to remove %s/loader/random-seed: %m",
+                                       FORMAT_FD_PATH(esp_fd, &buf));
+
+        log_info("Removed %s/random-seed.", FORMAT_FD_PATH(loader_fd, &buf));
```

Comment:
Dropped removal of files that are not written to by systemd-firstboot except for /etc/machine-info since it doesn't seem too far fetched that systemd-firstboot might write stuff there as well in the future.